### ISSUE TITLE
Is it legal that constructors of external objects take iteration variables of the initial system as parameters?

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#1705]
**Reported by beutlich on 24 Apr 2015 14:38 UTC**
This is the revised version of #1673: Is it legal that constructors of external objects take iteration variables of the initial system as parameters?

The example package ExtObjInitTest.mo contains two models V1 and V2 that compute the parameter/variable x in the initial system. It is then passed to the constructor of the external object VectorTable. It must be copied to the include directories of the C compiler such that it can be found.

The models simulation correctly if results of V1.y and V2.y give a straight line (starting at zero) with slope 0.73908513321515.

The package needs the external source [VectorTable.c](https://raw.githubusercontent.com/tbeu/ModelicaBook/fix-interpolateVectorTable/ModelicaByExample/Functions/Interpolation/source/VectorTable.c) (kindly copied from Mike Tiller's ModelicaByExample library).

```mo
within;
package ExtObjInitTest
  class VectorTable "A vector table implemented as an ExternalObject"
    extends ExternalObject;
    function constructor
      input Real ybar[:,2];
      output VectorTable table;
      external "C" table=createVectorTable(ybar, size(ybar,1)) annotation(Include="#include \"VectorTable.c\"");
    end constructor;
  
    function destructor "Release storage"
      input VectorTable table;
      external "C" destroyVectorTable(table) annotation(Include="#include \"VectorTable.c\"");
    end destructor;
  end VectorTable;

  function interpol
    "Interpolate a function defined by a vector using an ExternalObject"
    input VectorTable table;
    input Real x;
    output Real y;
    external "C" y = interpolateVectorTable(table, x) annotation(Include="#include \"VectorTable.c\"");
  end interpol;
  
  model V1 "Correct, if y at time=1 is 0.73908513321515"
    parameter Real x(fixed=false, start=0.25);
    Real y;
    protected
      VectorTable v=VectorTable({{0,0},{1,x}});
    initial equation
      x = cos(x);
    equation
      y = interpol(v, time);
    annotation(experiment(StopTime=1));
  end V1;

  model V2 "Correct, if y at time=1 is 0.73908513321515"
    Real x(start=0.25);
    Real y;
    protected
      VectorTable v=VectorTable({{0,0},{1,x}});
    equation
      when initial() then
        x = cos(x);
      end when;
      y = interpol(v, time);
    annotation(experiment(StopTime=1));
  end V2;
end ExtObjInitTest;
```

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1705

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 24 Apr 2015 14:43 UTC**
Should read: VectorTable.c must be copied to the include directories of the C compiler such that it can be found.

**Comment by modelica-trac-importer** [comment]
**Comment by thiele on 24 Apr 2015 16:29 UTC**
I don't think that the specification forbids that, hence seems valid to me.

However, it is currently not supported in OpenModelica and omc gives a warning about that missing support for the V2 example and fails with some strange error message for the V1 example.

I tried to find similar concepts in Modelica that could give some additional indication whether this should be allowed. I tried an example using a record constructor instead of an external object constructor. It simulates using omc.

```modelica
package InitialIteration
  record A
    Real x = -0.25;
    Real y;
  end A;

  model V1 "Should end at about 0.74 at t=1"
    parameter Real x(start=-0.25, fixed = false);
    Real y(start=0, fixed=true);
    InitialIteration.A a=A(x, y);
  initial equation 
    x = cos(x);
  equation 
    der(y) = a.x;
  end V1;

  model V2 "Should end at about 0.74 at t=1"
    Real x(start=-0.25, fixed = false);
    Real y(start=0, fixed=true);
    InitialIteration.A a= A(x,y);
  equation 
    when initial() then
      x = cos(x);
    end when;
    der(y) = a.x;
  end V2;
end InitialIteration;
```

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 14 May 2015 20:00 UTC**
[VectorTable.c](https://raw.githubusercontent.com/xogeny/ModelicaBook/master/ModelicaByExample/Functions/Interpolation/source/VectorTable.c) was fixed and merged and can be used from [Mike's repo](https://github.com/xogeny/ModelicaBook/blob/master/ModelicaByExample/Functions/Interpolation/source/VectorTable.c) now.

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 13 Jun 2016 18:06 UTC**
I think it is at least against the spirit of the specification if not plainly illegal:
Section 12.9.7 puts great emphasis on the fact that the constructor is evaluated exactly once:

> The constructor function is called exactly once before the first use of the object.

> It is not legal to call explicitly the constructor and destructor functions.
 
> external object can only be bound in component declarations and neither modified later nor assigned to

IMO there are two (incompatible) ways to implement the desired feature:

1. One could extend the C-interface in a way that handles arguments of continuous-time variability special (i.e. instead of a float, VectorTable could get a pointer to the location where the tool will store the continuous-time value during integration). This is quite problematic since we do not have the syntax to distinguish variability at the function declaration site. Additionally there would be many different permutations for a given function to implement.

2. One could allow the (re-)construction of external objects during integration with the current value of a continuous time expression. This is problematic as well, since it would require precise specification of the construction/destruction/reconstruction interaction. Right now, it will probably lead to either memory leaks or an unsafe state.


However, I do not think that such a feature is truly necessary. In any case, the model can be rewritten by supplying the continuous time values in the equation:

```mo
y = interpol(v, time, {1,x});
```

This should have the same efficiency and expressive power, so I do not see any reason to complicate the external interface in this regard.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 23 Jun 2016 19:54 UTC**
Thank you, Christoph. I do believe this is the expected answer for my question.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 23 Jun 2016 21:50 UTC**
I am closing this issue as it is dealt with in follow-up ticket #1907.

**Comment by modelica-trac-importer** [comment]
**Comment by M&#x2e;Arzt on 4 Jul 2016 16:27 UTC**
Both Tickets discuss different questions.

This Ticket: Which parameters are allowed to be passed to the constructor of an external object?

Ticket #1907: Is it feasible for the initialization of an external object, to involve function calls and initial algorithms besides the constructor?

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 5 Aug 2016 08:08 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 5 Aug 2016 08:41 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 13 Sep 2016 13:07 UTC**
Solved by reconstructor of #2037
